### PR TITLE
[FLINK-17307] Add collector to deserialize in KafkaDeserializationSchema

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR adds a way to emit multiple records from
KafkaDeserializationSchema. This is possible through a collector, which
will buffer deserialized records in a queue and then emit all records
atomically. The queue is reused for all incoming Kafka records to
minimize creating new objects on the hot path.

## Verifying this change
All existing kafka tests should pass. 
Added:
* org.apache.flink.streaming.connectors.kafka.KafkaConsumerTestBase#runCollectingSchemaTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
cc @aljoscha

**Comment by dawidwys** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTest.java [hunk 1]
```diff
@@ -108,13 +112,13 @@ public void testSkipCorruptedRecord() throws Exception {
 
 		final KafkaTopicPartitionState<Object> partitionStateHolder = fetcher.subscribedPartitionStates().get(0);
 
-		fetcher.emitRecord(1L, partitionStateHolder, 1L);
-		fetcher.emitRecord(2L, partitionStateHolder, 2L);
+		emitRecord(fetcher, 1L, partitionStateHolder, 1L);
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTest.java:115]: The changes in this file are purely an orthogonal refactoring, right? Could you put these in a separate commit, along with the removal of the `emitRecord()` method on the fetcher that is only used in tests?

File: flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java [hunk 2]
```diff
@@ -228,4 +226,33 @@ protected void doCommitInternalOffsetsToKafka(
...
 	}
+
+	private class KafkaCollector implements Collector<T> {
+		private final Queue<T> records = new ArrayDeque<>();
+
+		private boolean endOfStreamSignalled = false;
+
+		@Override
+		public void collect(T record) {
+			// do not emit subsequent elements if the end of the stream reached
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:237]: So we're not emitting all the records of the batch after one of them signals "end of stream"? I recall that an initial version still emitted all records.

**Comment by aljoscha** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java (hunk 2 above)
Comment (dawidwys) [reply flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:237]: Actually, that is the behaviour I had from the very beginning. Do you think it makes more sense to emit all but the end of the stream record?

Side note. I find this method very confusing. I just realized that there is no cross partition alignment on this method. The whole task will be brought down if any of the assigned partitions signals the end of stream, irrespective of the state of the remaining partitions. Honestly I'd be in favour of dropping this method at some point in the future.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTest.java (hunk 1 above)
Comment (dawidwys) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTest.java:115]: There are two related changes:
* I changed a bit the way no results are emitted (null before/empty queue now)
* I changed one test case slightly to emit multiple results in a single `emitRecordsWithTimestamps`

I think after extracting the refactoring to a separate commit it is visible more clearly.

File: flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java (hunk 2 above)
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:237]: Oh boy, I guess the interplay of this end record and multiple partitions was not considered... 🙈

Comment (dawidwys) [reply flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:237]: It had no additional implications on this PR as it has those problems already.

**Comment by begginghard** [comment]
@flinkbot run azure
